File: agentic-crawler/src/utils/htmlCleaner.ts
import * as cheerio from 'cheerio';

const NOISE_SELECTORS = [
  'script',
  'style',
  'svg',
  'img',
  'noscript',
  'iframe',
  'head',
  'header',
  'footer',
  'nav',
  '[aria-hidden="true"]',
  '.cookie-banner',
  '.advertisement',
  '.sidebar',
].join(', ');

/**
 * Strips noise elements and returns the main content HTML.
 * Keeps the HTML structure so the LLM can infer Q&A nesting.
 */
export function cleanHtml(rawHtml: string): string {
  const $ = cheerio.load(rawHtml);

  $(NOISE_SELECTORS).remove();

  // Prefer the main content container; fall back to body
  const main = $('main, article, [role="main"], #content, .content, #main').first();
  const content = main.length ? main.html() : $('body').html();

  return (content ?? '')
    .replace(/<!--[\s\S]*?-->/g, '')  // strip HTML comments
    .replace(/[ \t]{2,}/g, ' ')       // collapse inline whitespace
    .replace(/\n{3,}/g, '\n\n')       // collapse blank lines
    .trim();
}

/**
 * Rough token estimate: 1 token ≈ 4 characters.
 * Used to guard against sending oversized payloads to the LLM.
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

File: agentic-crawler-py/src/utils/html_cleaner.py
"""HTML cleaning utilities — BeautifulSoup4 port of htmlCleaner.ts."""

from __future__ import annotations

import re

from bs4 import BeautifulSoup

NOISE_SELECTORS = [
    "script",
    "style",
    "svg",
    "img",
    "noscript",
    "iframe",
    "head",
    "header",
    "footer",
    "nav",
    '[aria-hidden="true"]',
    ".cookie-banner",
    ".advertisement",
    ".sidebar",
]


def clean_html(raw_html: str) -> str:
    """Strip noise elements and return the main content HTML."""
    soup = BeautifulSoup(raw_html, "lxml")

    for selector in NOISE_SELECTORS:
        for tag in soup.select(selector):
            tag.decompose()

    # Prefer main content container; fall back to body
    main = soup.select_one("main, article, [role='main'], #content, .content, #main")
    content_tag = main or soup.body
    content = str(content_tag) if content_tag else ""

    # Strip HTML comments
    content = re.sub(r"<!--[\s\S]*?-->", "", content)
    # Collapse inline whitespace
    content = re.sub(r"[ \t]{2,}", " ", content)
    # Collapse blank lines
    content = re.sub(r"\n{3,}", "\n\n", content)
    return content.strip()


def estimate_tokens(text: str) -> int:
    """Rough token estimate: 1 token ≈ 4 characters."""
    return (len(text) + 3) // 4
